### PR TITLE
fix(ui): show a 404 for an unknown template id instead of a blank page (#496)

### LINKED ISSUE #496: Update Data Type for Template ID
Currently, accessing URLs with reaction IDs or dataset IDs 

Image

But for incorrect Template ID we have empty page

Template http://localhost:5173/templates/111111111111

Image

Please update the data type to support longer Template IDs

Enchantments is related with issue 
https://github.com/open-reaction-database/ord-app/issues/445

### PR DESCRIPTION
**Pull request by skearnes**
## Summary

Visiting `/templates/<unknown-id>` rendered a **blank page** — `TemplatePage` guards its content on `{template && ...}` and silently renders nothing when the template isn't in the store. Because templates are fetched in bulk via `getAllTemplates`, a missing id is indistinguishable from "still loading" without an explicit signal.

**Fix:** add an `areTemplatesLoaded` flag to the templates slice (flips to `true` once the full list is fetched) + a `selectAreTemplatesLoaded` selector. `TemplatePage` now renders the shared `NotFoundPage` (404) when the list **has loaded** and the id is unknown, while staying quiet during the initial load so it doesn't **flash** a 404.

### Scope notes
- Long template ids already work — the page keys templates by string (`template_<id>`), so there's no numeric overflow and no "data type" change needed.
- The issue's secondary ask (datasets returning **404 instead of 403** for unknown ids) is backend work tracked under #446 and is **out of scope** here.

## Test plan
- [x] `tsc -b` clean; `prettier` / `eslint` clean; **689 tests pass**
- [x] New tests (8 across reducer/selector/page):
  - reducer: `areTemplatesLoaded` is `false` initially, `true` after `getAllTemplates` success
  - selector: `selectAreTemplatesLoaded` reflects the flag
  - page: renders 404 when loaded + unknown id; **no** 404 while still loading; renders content when the template exists
- [ ] Not booted on the stack for this one — the page-render test directly covers all three branches (404 / loading / present); a quick manual nav to `/templates/999999` would be a nice final eyeball before merge.

Closes #496 (template half).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR fixes a blank-page bug where navigating to `/templates/<unknown-id>` silently rendered nothing because `TemplatePage` guarded its content on `{template && ...}` with no way to distinguish "still loading" from "genuinely missing."

- Adds an `areTemplatesLoaded` boolean to the templates Redux slice (settles to `true` on both `getAllTemplates` success and failure) and a matching `selectAreTemplatesLoaded` selector.
- `TemplatePage` now short-circuits to `NotFoundPage` (404) when the list has settled and the id is absent, while keeping the existing `PageContainer` render during the initial load to avoid flashing a 404 prematurely.
- All three branches (404 / loading / present) are covered by new unit tests alongside reducer and selector tests, including the failure-path case addressed during review.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is narrow, well-tested, and the failure-path edge case flagged in review has been addressed.

The fix is confined to three small files: a new sub-reducer with a single `addMatcher`, a one-line selector, and an early-return guard in the page component. The reducer correctly settles the flag on both success and failure, and the three-branch test matrix (404 / loading / present) gives good coverage of the added logic. No existing behaviour is altered for users who already have templates loaded.

No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| ui/src/pages/TemplatePage/TemplatePage.tsx | Adds the `areTemplatesLoaded` guard before the main render: shows `NotFoundPage` (404) when the list has settled and the id is absent; keeps rendering `PageContainer` while still loading to avoid flashing a 404. |
| ui/src/store/entities/templates/templates.reducer.ts | Adds `areTemplatesLoaded` reducer (flips to `true` on both success and failure of `getAllTemplates`) and includes it in `combineReducers`; correctly handles network failure so the 404 path is reachable even when the initial fetch errors out. |
| ui/src/store/entities/templates/templates.selectors.ts | Adds `selectAreTemplatesLoaded` using the existing `buildSelector` factory — minimal and consistent with surrounding selectors. |
| ui/src/pages/TemplatePage/TemplatePage.test.tsx | New test file covering all three branches of the guard (404 when loaded + unknown, no flash while loading, content when template exists) using hoisted mocks for clean isolation. |
| ui/src/store/entities/templates/templates.reducer.test.ts | Adds tests for `areTemplatesLoaded` covering initial state, success case, and the failure case added after the previous review thread. |
| ui/src/store/entities/templates/templates.selectors.test.ts | Adds a focused test for `selectAreTemplatesLoaded` confirming it reflects both `false` and `true` states from the store shape. |

</details>

</details>

<sub>Reviews (2): Last reviewed commit: ["fix(ui): settle areTemplatesLoaded on fe..."](https://github.com/open-reaction-database/ord-app/commit/b63328d0b4a15cc3508c1c6303a36f7c69bd157d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38757774)</sub>

### CODE REVIEW

File: ui/src/store/entities/templates/templates.reducer.ts [hunk 1]
```diff
@@ -53,7 +53,15 @@ const isTemplateCreating = createReducer<boolean>(false, builder => {
   );
 });
 
+// Whether the full template list has been fetched. Lets the template page tell "still loading"
+// apart from "no such template" so it can show a 404 for a missing id rather than a blank page,
+// without flashing the 404 during the initial load. (#496)
+const areTemplatesLoaded = createReducer<boolean>(false, builder => {
+  builder.addCase(getAllTemplatesActions.success, () => true);
+});
```
[the remark this answers is not in the document]
Comment (skearnes) [reply ui/src/store/entities/templates/templates.reducer.ts:62]: Good catch — fixed. `areTemplatesLoaded` now settles to `true` on `getAllTemplatesActions.failure` as well as success (treating it as "fetch settled"), so a failed fetch still reaches the 404 path instead of leaving a permanently blank page. Added a reducer test for the failure case.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
